components/sock_utils/src/socketpair.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * SPDX-FileCopyrightText: 2024 Espressif Systems (Shanghai) CO LTD
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <fcntl.h>
+#include "lwip/sockets.h"
+#include "socketpair.h"
+#include "esp_check.h"
+
+#define INVALID_SOCKET (-1)
+
+static const char *TAG = "socket_helpers";
+
+static int set_nonblocking(int sock)
+{
+    int opt;
+    opt = fcntl(sock, F_GETFL, 0);
+    if (opt == -1) {
+        return -1;
+    }
+    if (fcntl(sock, F_SETFL, opt | O_NONBLOCK) == -1) {
+        return -1;
+    }
+    return 0;
+}
+
+int socketpair(int domain, int type, int protocol, int sv[2])
+{
+    struct sockaddr_storage ss;
+    struct sockaddr_in *sa = (struct sockaddr_in *)&ss;
+    socklen_t ss_len = sizeof(struct sockaddr_in);
+    int fd1 = INVALID_SOCKET;
+    int fd2 = INVALID_SOCKET;
+    int listenfd = INVALID_SOCKET;
+    int ret = 0; // Success
+
+    sa->sin_family = AF_INET;
+    sa->sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+    sa->sin_port = 0;
+    listenfd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+    ESP_GOTO_ON_FALSE(listenfd != INVALID_SOCKET, -1, err, TAG, "Cannot create listening socket");
+    ESP_GOTO_ON_FALSE(listen(listenfd, 1) == 0, -1, err, TAG, "Failed to listen");
+
+    memset(&ss, 0, sizeof(ss));
+    ss_len = sizeof(ss);
+    ESP_GOTO_ON_FALSE(getsockname(listenfd, (struct sockaddr *)&ss, &ss_len) >= 0, -1, err, TAG, "getsockname failed");
+
+    sa->sin_family = AF_INET;
+    sa->sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+
+    fd1 = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+    ESP_GOTO_ON_FALSE(fd1 != INVALID_SOCKET, -1, err, TAG, "Cannot create read side socket");
+    ESP_GOTO_ON_FALSE(set_nonblocking(fd1) == 0, -1, err, TAG, "Failed to set socket to nonblocking mode");
+    if (connect(fd1, (struct sockaddr *)&ss, ss_len) < 0) {
+        ESP_GOTO_ON_FALSE(errno == EINPROGRESS || errno == EWOULDBLOCK, -1, err, TAG, "Failed to connect fd1");
```

Comment:
Here only two errors are handled. Can we handle any error?

Reply:
we handle all other errors, but it's okay if the errno is either EINPROGRESS or EWOULDBLOCK

we can log the actual errno.

Reply:
I switched to blocking mode, so it's not needed anymore.
removed completely.